# CSE3000 Research Project - Bram Pronk - Group 21
### Assessing How Variational Auto-Encoders Can Combine Information From Multiple Data Sources in Cancer Cells

[![Python Version](https://img.shields.io/static/v1.svg?label=minimal_python_version&message=3.8.8&color=blue)](https://www.python.org/downloads)

## Description of the Research
The main question this research aims to answer is: "How can Variational Auto-Encoders be used to combine information from multiple data sources in cancer cells?" One of the defining features of VAE’s is their ability to achieve dimensionality reduction such that their latent space is not only representative of original data, but also viable for generation of new data [5]. The high-dimensional data set [8] consists of samples from ten thousand tumors of 33 of the most prevalent cancers. To each sample, RNA-Seq [9] has been applied. This means for each sample we can measure gene expression of RNA, changes in the DNA sequence, changes in the number of chromosomes and chemical modifications of DNA. The results gathered here are thus in whole or part based upon data generated by The Cancer Genome Atlas (TCGA) Research Network [3]. This project aims to provide an argument for using VAE’s in the process of finding treatment that fits best to each individual patient.

## Getting Started
<!---

This section should contain installation, testing, and running instructions for people who want to get started with the project. 

- These instructions should work on a clean system.
- These instructions should work without having to install an IDE.
- You can specify that the user should have a certain operating system.

--->
This project will compare a linear Multi-Omics Factor Analysis (MOFA) with two proposed models for VAE's that span multiple data modalities (MVAE), namely Mixture-of-Experts (MoE) and Product-of-Experts (PoE). 
To that end, this repository will have three branches for each model; ```mofa```, ```moe``` and ```moe```. 
Since each model will be a fork from an existing repository that is adjusted to fit TCGA's data, environment setup will differ. To that end, instructions for installation and running will be provided in the README.md of each branch.

## Authors
This is the personal repository of

    - Bram Pronk            (4613066) i.b.pronk@student.tudelft.nl

My colleagues in Research Group 21:

     - Armin Korkic         (4713052) a.korkic@student.tudelft.nl
     - Boris van Groeningen (4719875) b.vangroeningen@student.tudelft.nl
     - Ivo Kroskinski       (4684958) i.s.kroskinski@student.tudelft.nl
     - Raymond d'Anjou      (4688619) r.danjou@student.tudelft.nl

Research Group 21 is guided by:
    
    - Marcel Reinders (Responsible Professor)
    - Stavros Makrodimitris, Tamim Abdelaal, Mohammed Charrout, Mostafa elTager (Supervisors)

## License
TBD.





